XenAdmin/Diagnostics/Checks/RebootPendingOnMasterCheck.cs

```diff
@@ -0,0 +1,127 @@
+ * SUCH DAMAGE.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using XenAdmin.Core;
+using XenAdmin.Diagnostics.Hotfixing;
+using XenAdmin.Diagnostics.Problems;
+using XenAdmin.Diagnostics.Problems.HostProblem;
+using XenAdmin.Diagnostics.Problems.PoolProblem;
+using XenAPI;
+
+namespace XenAdmin.Diagnostics.Checks
+{
+    class RestartHostOrToolstackPendingOnMasterCheck : Check
+    {
+        public string UpdateUuid { get; private set; }
+        private readonly Pool pool;
+
+        public RestartHostOrToolstackPendingOnMasterCheck(Pool pool, string updateUuid)
+            : base(Helpers.GetMaster(pool.Connection))
+        {
+            this.pool = pool;
+            this.UpdateUuid = updateUuid;
+        }
+
+        protected override Problem RunCheck()
+        {
+            if (!Host.IsLive())
+                return new HostNotLiveWarning(this, Host);
+            
+            var elyOrGreater = Helpers.ElyOrGreater(Host);
+
+            double bootTime = Host.BootTime();
+            double agentStart = Host.AgentStartTime();
+
+            if (bootTime == 0.0 || agentStart == 0.0)
+                return null; //fine
+
+            //check reboot
+            if (elyOrGreater)
+            {
+                foreach (var updateRef in Host.updates_requiring_reboot)
+                {
+                    var update = Host.Connection.Resolve(updateRef);
+
+                    if (string.IsNullOrEmpty(UpdateUuid) || //automated mode, any update
+                        string.Equals(update.uuid, UpdateUuid, System.StringComparison.InvariantCultureIgnoreCase)) //normal mode the given update
+                    {
+                        return new MasterIsPendingRestartHostProblem(this, pool);
+                    }
+                }
+            }
+            else
+            {
+                var hostRestartRequiredPatches = Host.AppliedPatches().Where(p => p.after_apply_guidance.Contains(after_apply_guidance.restartHost) && ((double)Util.ToUnixTime(p.AppliedOn(Host)) > agentStart));
+
+                foreach (Pool_patch patch in hostRestartRequiredPatches)
+                {
+                    if (string.IsNullOrEmpty(UpdateUuid) //automated mode, any update
+                            || string.Equals(patch.uuid, UpdateUuid, System.StringComparison.InvariantCultureIgnoreCase)) //normal mode the given update
+                    {
+                        return new MasterIsPendingRestartHostProblem(this, pool);
+                    }
+                }
+            }
+
+            //check toolstack restart
+            var toolstackRestartRequiredPatches = Host.AppliedPatches().Where(p => p.after_apply_guidance.Contains(after_apply_guidance.restartXAPI) && ((double)Util.ToUnixTime(p.AppliedOn(Host)) > agentStart));
+            foreach (Pool_patch patch in toolstackRestartRequiredPatches)
+            {
+                if (string.IsNullOrEmpty(UpdateUuid)) //automated mode

```

Comment:
Actually, this check will never happen, because it returns already. I don't know if we could put the following code on line 60 to avoid repeated checks:
    if (string.IsNullOrEmpty(UpdateUuid))
        return new MasterIsPendingRestartHostProblem(this, pool);

Reply:
This is there for a reason, we check host-restart requirements first, and only then the toolstack-restart ones. I think we should keep it as is.